Add empty rust crates for protobufs

nick-mobilecoin opened the pull request.

nick-mobilecoin committed:
Add empty rust crates for protobufs

nick-mobilecoin commented:
Current dependencies on/for this PR:
* main
  * **PR #4** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/protobufs/4" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈
    * **PR #6** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/protobufs/6" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #7** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/protobufs/7" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #9** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/protobufs/9" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/protobufs/4?utm_source=stack-comment).

nick-mobilecoin commented on crates/README.md:
```diff
@@ -0,0 +1,12 @@
+
+This is composed of 3 crates:
+
+1. `mc-probufs-tonic`: A crate providing a
```
I'm up for suggestions on these crate names, I had `mc-grpc-` but it looked kind of weird for `mc-grpc-grpcio`.  `mc-proto-` or `mc-protobufs-` also seem reasonable.

nick-mobilecoin replied:
naming is being addressed from initial review

nick-mobilecoin commented on crates/.gitignore:
```diff
@@ -0,0 +1,10 @@
+# Generated by Cargo
```
Generally I'm not fond a of a `crates` directory, seen that in some repos that look to be "only" rust.
However since the primary purpose of this repo is to house the proto files and as a convenience we're providing rust crates it seemed reasonable

jcape replied:
I think there's two options here:
1. Have a `rust` directory (fka `crates`) that all this stuff lives under.
```
mobilecoinfoundation/
    attestation/
        v1/
            foo.protos
rust/
    attestation/
        tonic/
            mod v1
        grpcio/
            mod v1
        messages/
            mod v1
```
2. Interleave this with the protobuf files, e.g.:
```
mobilecoinfoundation/
    attestation/
        v1/
            foo.protos
        tonic/
            mod v1
        grpcio/
            mod v1
        messages/
            mod v1
```

I kind of prefer the latter, but that's strictly out of a desire to avoid parallel filesystem trees...

nick-mobilecoin replied:
Did the interleaving, I verified buf.build only uploads protos and not other dirs/files.
The generated files will be `mobilecoin.attestation.v1.rs` so I think that should work

nick-mobilecoin commented on crates/deny.toml:
```diff
@@ -0,0 +1,55 @@
+allow = []
+deny = [
+    # https://github.com/briansmith/ring/issues/774
+    { name = "ring" },
```
This is copy pasta from our other deny, but I'm pretty sure tonic and grpc use `ring` so we may need to omit as we move along

nick-mobilecoin commented on crates/messages/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "mc-probufs-messages"
```
up for suggestions on the common message format crate, it could be that `types` is a better name :shrug:

jcape replied:
I actually like messages, TBH, I think it's a great pattern to disambiguate from the mc-foo-types being in rust.

nick-mobilecoin replied:
renamed back to messages

jcape reviewed (changes requested):
Ideally there would be a different crate for each package. i.e. the crate would be, like, `mc-attestation-types-grpcio` or `mc-protobufs-attestation-types-grpcio`.

The thinking is: these crates are the place to host type conversions to "business logic" rust types, which means they will need to depend on the `foo-types`-style crates, and I'd love to avoid having a huge dependency tree behind using a subset of these protobufs.

nick-mobilecoin commented:
> Ideally there would be a different crate for each package. i.e. the crate would be, like, `mc-attestation-types-grpcio` or `mc-protobufs-attestation-types-grpcio`.
> 
> The thinking is: these crates are the place to host type conversions to "business logic" rust types, which means they will need to depend on the `foo-types`-style crates, and I'd love to avoid having a huge dependency tree behind using a subset of these protobufs.

Ensuring we're on the same page
```mermaid
graph LR
    A[mc-attestation-tonic] --> B[mc-attestation-protobuf-types]
    C[mc-attestation-grpcio] --> B
    B --> D[mc-attestation-types]
```
- The `mc-attestation-tonic` and `mc-attestation-grpcio` are the interfaces that implement grpc attestation. I kind of want to have grpc in the name but it's confusing with grpcio. Following their names to the respective crate implementations will hopefully be clear enough.
- The `mc-attestation-protobuf-types` are the types used in the grpc traffic. Note, using prost and diffing the generated content it seems we only need one common protobuf type for both tonic and grpcio.  We may consider `mc-attestation-prost-types` too.
- The `mc-attestation-types` contain the raw rust attestation types. These types should be isolated by the chosen transport protocol an thus why this crate intentionally does *not* depend on `mc-attestation-protobuf-types`. i.e. if we decide to serialize in JSON or XML this crate should not need to change.

nick-mobilecoin committed:
Narrow crate scope to `attestation`

Previously the crates were named to contain all future protobuf types
and services, now the crates have been named to only support
`attestation` with the intent to create dedicated crates for other gRPC
communications.

nick-mobilecoin committed:
Move rust crates down near the protobufs

The rust crates are interleaved in the protobuf directories. This
prevents mirroring the directory structure in another sub-directory

NotGyro reviewed (approved):
Looks good to me!
